src/Juvix/Frontend/Parser.hs

```diff
@@ -0,0 +1,807 @@
+{-# LANGUAGE ApplicativeDo #-}
+
+-- |
+-- - The front end parser for the Juvix Programming language
+--
+-- - Parsers with S at the end, eat the spaces at the end of the parse
+--
+-- - Parsers with SN at the end, eats the spaces and new lines at the
+--   end of the parse
+module Juvix.Frontend.Parser where
+
+import Data.Attoparsec.ByteString hiding (match)
+import qualified Data.Attoparsec.Expr as Expr
+import qualified Data.ByteString as ByteString
+import qualified Data.ByteString.Char8 as Char8
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Data.Set as Set
+import qualified Data.Text.Encoding as Encoding
+import qualified Juvix.Core.Usage as Usage
+import qualified Juvix.Frontend.Lexer as Lexer
+import qualified Juvix.Frontend.Types as Types
+import Juvix.Library hiding (guard, maybe, option, product, sum, take, takeWhile, try)
+import Prelude (fail)
+
+--------------------------------------------------------------------------------
+-- Top Level
+--------------------------------------------------------------------------------
+
+topLevel ∷ Parser Types.TopLevel
+topLevel =
+  Types.Type <$> typeP
+    <|> Types.ModuleOpen <$> moduleOpen
+    <|> Types.Module <$> module'
+    <|> Types.Function <$> function
+    <|> Types.Signature <$> signature'
+
+expressionGen' ∷ Parser Types.Expression → Parser Types.Expression
+expressionGen' p =
+  Types.Cond <$> cond
+    <|> Types.Let <$> let'
+    <|> Types.Match <$> match
+    <|> Types.OpenExpr <$> moduleOpenExpr
+    <|> Types.Block <$> block
+    <|> Types.Lambda <$> lam
+    <|> Types.ExpRecord <$> expRecord
+    <|> Types.Application <$> try application
+    <|> try p
+    <|> Types.Constant <$> constant
+    <|> Types.Name <$> prefixSymbolDot
+    <|> parens (expressionGen p)
+
+expressionGen ∷ Parser Types.Expression → Parser Types.Expression
+expressionGen p =
+  Expr.buildExpressionParser tableExp (spaceLiner (expressionGen' p))
+
+expression' ∷ Parser Types.Expression
+expression' = expressionGen (fail "")
+
+expression ∷ Parser Types.Expression
+expression = expressionGen (Types.Do <$> do')
+
+usage ∷ Parser Types.Expression
+usage = string "u#" *> expression
```

Comment:
Do we really need this prefix to all usage expressions?

Reply:
I don't use this at all, this can be removed. Arrows and signatures already have it properly built in

Reply:
Nice, let's remove it then.